File: src/solution7/bread/TravelRoute.java
package solution7.bread;
import java.util.*;

public class TravelRoute {
    List<String> list = new ArrayList<>();
	static String route = "";
	static boolean[] visit;
	
	private void dfs(String[][] tickets, String end, int cnt) {
		route += end + ",";
		
		if(cnt == tickets.length) {
			list.add(route);
            return;
		}
		
		for(int i = 0; i < tickets.length; i++) {
			String s = tickets[i][0];
            String e = tickets[i][1];
            
			if(s.equals(end) && !visit[i]) {
				visit[i] = true;
				dfs(tickets, e, cnt + 1);
				visit[i] = false;
                route = route.substring(0, route.length()-4);
			}
		}
	}
	
	public String[] solution(String[][] tickets) {
		for(int i = 0; i < tickets.length; i++) {
			visit = new boolean[tickets.length];
			String start = tickets[i][0];
            String end = tickets[i][1];
			
			if(start.equals("ICN")) {
				route = start + ",";
                visit[i] = true; 
				dfs(tickets, end, 1);
			}
		}
		
		Collections.sort(list);
		String[] answer = list.get(0).split(",");

		return answer;
	}

    public static void main(String[] args) {
		// TODO Auto-generated method stub
        String[][] tickets = {{"ICN", "JFK"}, {"HND", "IAD"}, {"JFK", "HND"}};
        
		TravelRoute temp = new TravelRoute();
		
		System.out.println(temp.solution(tickets));
	}
}

File: src/solution7/nana/travel-route.js
function getAdjacent(tickets) {
  const adj = {};
  for (let [src, dst] of tickets) {
    if (adj[src]) {
      adj[src].push(dst);
    } else {
      adj[src] = [dst];
    }
  }
  Object.keys(adj).forEach((key) => {
    adj[key].sort((a, b) => {
      if (a < b) return 1;
      if (a > b) return -1;
      return 0;
    });
  });
  return adj;
}

function dfs(ans, adj, src, cnt) {
  ans.push(src);
  if (cnt === 0) return true; // 티켓 다 씀
  if (adj[src]) {
    for (let i = 0; i < adj[src].length; i++) {
      let dist = adj[src].pop();
      if (dfs(ans, adj, dist, cnt - 1)) return true;
      adj[src].unshift(dist);
    }
  }
  ans.pop();
  return false;
}

function solution(tickets) {
  var answer = [];
  const adj = getAdjacent(tickets);
  dfs(answer, adj, 'ICN', tickets.length);
  return answer;
}

console.log(
  solution([
    ['ICN', 'JFK'],
    ['HND', 'IAD'],
    ['JFK', 'HND'],
  ])
);
console.log(
  solution([
    ['ICN', 'SFO'],
    ['ICN', 'ATL'],
    ['SFO', 'ATL'],
    ['ATL', 'ICN'],
    ['ATL', 'SFO'],
  ])
);
console.log(
  solution([
    ['ICN', 'COO'],
    ['ICN', 'BOO'],
    ['COO', 'ICN'],
  ])
);
